velox/common/caching/SsdFile.cpp

```diff
@@ -0,0 +1,412 @@
+      scoreSum += regionScore_[i];
+      candidates.push_back(i);
+    }
+  }
+  if (candidates.empty()) {
+    suspended_ = true;
+    return false;
+  }
+  // Sort by score to evict less read regions first.
+  std::sort(
+      candidates.begin(), candidates.end(), [&](int32_t left, int32_t right) {
+        return regionScore_[left] < regionScore_[right];
+      });
+  // Free up to 3 lowest score regions.
+  if (candidates.size() > 3) {
+    candidates.resize(3);
+  }
+  clearRegionEntriesLocked(candidates);
+  writableRegions_ = std::move(candidates);
+  suspended_ = false;
+  return true;
+}
+
+void SsdFile::clearRegionEntriesLocked(
+    const std::vector<int32_t>& regionIndices) {
+  // Remove all 'entries_' where the dependent points one of 'regionIndices'.
+  auto it = entries_.begin();
+  while (it != entries_.end()) {
+    auto region = regionIndex(it->second.offset());
+    if (std::find(regionIndices.begin(), regionIndices.end(), region) !=
+        regionIndices.end()) {
+      it = entries_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  for (auto region : regionIndices) {
+    // While the region is being filled it may get score from
+    // hits. When it is full, it will get a score boost to be a little
+    // ahead of the best.
+    regionScore_[region] = 0;
+    regionSize_[region] = 0;
+  }
+}
+
+void SsdFile::write(std::vector<CachePin>& pins) {
+  std::sort(pins.begin(), pins.end());
+  uint64_t total = 0;
+  for (auto& pin : pins) {
+    total += pin.entry()->size();
+  }
+  int32_t storeIndex = 0;
+  while (storeIndex < pins.size()) {
+    auto [offset, available] = getSpace(pins, storeIndex);
+    if (!available) {
+      // No space can be reclaimed. The pins are freed when the caller is freed.
+      return;
+    }
+    int32_t numWritten = 0;
+    int32_t bytes = 0;
+    std::vector<iovec> iovecs;
+    for (auto i = storeIndex; i < pins.size(); ++i) {
+      auto entrySize = pins[i].entry()->size();
+      if (bytes + entrySize > available) {
+        break;
+      }
+      addEntryToIovecs(*pins[i].entry(), iovecs);
+      bytes += entrySize;
+      ++numWritten;
+    }
+    VELOX_CHECK_GE(fileSize_, offset + bytes);
+    auto rc = folly::pwritev(fd_, iovecs.data(), iovecs.size(), offset);
+    if (rc != bytes) {
+      LOG(ERROR) << "Failed to write to SSD " << errno;
+      // If the write fails we return without adding the pins to the cache. The
+      // entries are unchanged.
+      return;
+    }
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      for (auto i = storeIndex; i < storeIndex + numWritten; ++i) {
+        auto entry = pins[i].entry();
+        entry->setSsdFile(this, offset);
+        char first = entry->tinyData() ? entry->tinyData()[0]
+                                       : entry->data().runAt(0).data<char>()[0];
+        auto size = entry->size();
+        SsdKey key = {
+            entry->key().fileNum, static_cast<uint64_t>(entry->offset())};
```

Comment:
nit: use offset for entry->offset() for clarity

Reply:
Looks like this comment got lost. Repeating. "use offset for entry->offset() for clarity"